beluga_tutorial/src/main.cpp

```diff
@@ -0,0 +1,264 @@
+    resample["states"] = beluga::views::states(record.resample) | ranges::to<std::vector<double>>;
+    resample["weights"] = beluga::views::weights(record.resample) | ranges::to<std::vector<double>>;
+
+    auto estimation = node["simulation_records"][cycle]["estimation"];
+    estimation["mean"] = std::get<0>(record.estimation);
+    estimation["sd"] = std::get<1>(record.estimation);
+  }
+
+  std::ofstream fout(path, std::ios::trunc);
+  fout << node;
+  fout.close();
+}
+
+int main(int argc, char* argv[]) {
+  if (argc < 2) {
+    std::cerr << "Usage: " << argv[0] << "<parameters_path>\n";
+    return 1;
+  }
+
+  Parameters parameters;
+  load_parameters(argv[1], parameters);
+
+  SimulationRecord simulation_record;
+
+  LandmarkMapVector landmark_map;
+  load_landmark_map(argv[1], landmark_map);
+  simulation_record.landmark_map = landmark_map;
+
+  std::normal_distribution<double> initial_distribution(parameters.initial_position, parameters.initial_position_sd);
+  auto particles = beluga::views::sample(initial_distribution) |                  //
+                   ranges::views::transform(beluga::make_from_state<Particle>) |  //
+                   ranges::views::take_exactly(parameters.number_of_particles) |  //
+                   ranges::to<beluga::TupleVector>;
+
+  double current_position{parameters.initial_position};
+  for (unsigned int n = 0; n < parameters.number_of_cycles; ++n) {
+    RobotRecord robot_record;
+
+    current_position += parameters.velocity * parameters.dt;
+    robot_record.ground_truth = current_position;
+
+    if (current_position > parameters.map_size) {
+      break;
+    }
+
+    auto motion_model = [&](double particle_position, auto& gen) {
+      const double distance = parameters.velocity * parameters.dt;
+      std::normal_distribution<double> distribution(distance, parameters.translation_sd);
+      return particle_position + distribution(gen);
+    };
+
+    auto detections =
+        landmark_map |                                                                                       //
+        ranges::views::transform([&](double landmark) { return landmark - current_position; }) |             //
+        ranges::views::remove_if([&](double range) { return std::abs(range) > parameters.sensor_range; }) |  //
+        ranges::to<SensorData>;
+
+    auto sensor_model = [&](double particle_position) {
+      auto particle_detections =
+          landmark_map |                                                                             //
+          ranges::views::transform([&](double landmark) { return landmark - particle_position; }) |  //
+          ranges::to<SensorData>;
+
+      return parameters.min_particle_weight +
+             std::transform_reduce(
+                 detections.begin(), detections.end(), 1.0, std::multiplies<>{}, [&](double detection) {
+                   auto distances = particle_detections |  //
```

Comment:
nit: Maybe add `//` to prevent the weird indentation.

```suggestion
                   auto distances =         //
                     particle_detections |  //
```